Fabric.Authorization.API/Modules/UsersModule.cs

```diff
@@ -47,35 +57,56 @@ public class UsersModule : FabricModule<User>
+            Get("/{identityProvider}/{subjectId}/groups",
+                async _ => await GetUserGroups().ConfigureAwait(false), null,
                 "GetUserGroups");
         }
 
-        private async Task<dynamic> GetCurrentUserPermissions()
+        private async Task<dynamic> GetUserPermissions(dynamic param)
         {
             var userPermissionRequest = this.Bind<UserInfoRequest>();
             await SetDefaultRequest(userPermissionRequest);
             await CheckAccess(_clientService, userPermissionRequest.Grain, userPermissionRequest.SecurableItem,
                 AuthorizationReadClaim);
 
-            var subjectId = SubjectId;
-            var identityProvider = IdentityProvider;
-            var groups = await GetGroupsForAuthenticatedUser(subjectId, identityProvider).ConfigureAwait(false);
-
-            var permissions = await _permissionService.GetPermissionsForUser(
-                $"{subjectId}:{identityProvider}",
-                groups,
-                userPermissionRequest.Grain,
-                userPermissionRequest.SecurableItem);
+            var permissionResolutionResult = await new PermissionResolver(_roleService, _permissionService, Logger).Resolve(new PermissionResolutionRequest
+            {
+                SubjectId = param.subjectId,
+                IdentityProvider = param.identityProvider,
+                Grain = userPermissionRequest.Grain,
+                SecurableItem = userPermissionRequest.SecurableItem,
+                UserGroups = await GetGroupsForAuthenticatedUser(SubjectId, IdentityProvider)
+            });
 
             return new UserPermissionsApiModel
             {
                 RequestedGrain = userPermissionRequest.Grain,
                 RequestedSecurableItem = userPermissionRequest.SecurableItem,
-                Permissions = permissions
+                Permissions = permissionResolutionResult.AllowedPermissions.Except(permissionResolutionResult.DeniedPermissions).Select(p => p.Name)
             };
         }
 
+        private async Task<dynamic> GetCurrentUserPermissions()
+        {
+            var userPermissionRequest = this.Bind<UserInfoRequest>();
+            await SetDefaultRequest(userPermissionRequest);
+            await CheckAccess(_clientService, userPermissionRequest.Grain, userPermissionRequest.SecurableItem,
+                AuthorizationReadClaim);
+
+            var permissionResolutionResult = await new PermissionResolver(_roleService, _permissionService, Logger).Resolve(new PermissionResolutionRequest
+            {
+                SubjectId = SubjectId,
+                IdentityProvider = IdentityProvider,
+                Grain = userPermissionRequest.Grain,
+                SecurableItem = userPermissionRequest.SecurableItem,
+                UserGroups = await _userService.GetGroupsForUser(SubjectId, IdentityProvider)

```

Comment:
For the current users permissions, we need to make sure we get the groups from the access token too.